relay-quotas/src/redis.rs

```diff
@@ -125,6 +139,300 @@ impl std::ops::Deref for RedisQuota<'_> {
+    /// we try again to decrement the budget after syncing.
+    fn evaluate_quota(
+        &self,
+        client: &mut PooledClient,
+        quota: &RedisQuota,
+        quantity: usize,
+    ) -> Result<bool, GlobalRateLimitError> {
+        use GlobalRateLimitError as E;
+
+        match self.decrement_budget(quota, quantity) {
+            ok @ Ok(_) => ok,
+            err @ Err(E::Redis | E::PoisonedLock | E::LoopLimitExceeded) => err,
+            Err(E::BudgetEmpty | E::SlotExpired | E::KeyMissing) => {
+                self.redis_sync(client, quota, quantity)?;
+                self.decrement_budget(quota, quantity)
+            }
+        }
+    }
+
+    /// Attempts to decrement the budget by a certain amount.
+    ///
+    /// Returns `true` if items should be ratelimited.
+    fn decrement_budget(
+        &self,
+        quota: &RedisQuota,
+        quantity: usize,
+    ) -> Result<bool, GlobalRateLimitError> {
+        let current_slot = quota.slot() as usize;
+        let Some(limit) = quota.limit.map(|limit| limit as usize) else {
+            return Ok(false);
+        };
+
+        let map = self
+            .counters
+            .read()
+            .map_err(|_| GlobalRateLimitError::PoisonedLock)?;
+
+        let val = map
+            .get(&BudgetKeyRef::new(quota))
+            .ok_or(GlobalRateLimitError::KeyMissing)?;
+
+        let max_loop_iterations = 10;
+        for _ in 0..max_loop_iterations {
+            if current_slot != val.slot.load(Ordering::SeqCst) {
+                // Expired slot implies the the time window has passed, and we need to
+                // ask for a new budget for the new slot.
+                return Err(GlobalRateLimitError::SlotExpired);
+            }
+
+            let redis_count = val.last_seen_redis_value.load(Ordering::SeqCst);
+            let current_budget = val.budget.load(Ordering::SeqCst);
+
+            // Since the redis count was pre-incremented by the amount of budget we took, the actual
+            // total count have to subtract the current budget.
+            let total_count = redis_count.saturating_sub(current_budget);
+
+            if total_count + quantity > limit {
+                return Ok(true);
+            }
+
+            if current_budget < quantity {
+                return Err(GlobalRateLimitError::BudgetEmpty);
+            }
+
+            // When setting the new value, we must ensure that it hasn't already been
+            // decremented in another thread.
+            if val
+                .budget
+                .compare_exchange_weak(
+                    current_budget,
+                    current_budget - quantity,
+                    Ordering::SeqCst,
+                    Ordering::SeqCst,
+                )
+                .is_ok()
+            {
+                return Ok(false);
+            }
+        }
+
+        Err(GlobalRateLimitError::LoopLimitExceeded)
```

Comment:
hopefully never happens, I was just nervous to have a potentially infinite loop in prod